pandas/_libs/tslibs/timestamps.pyx

```diff
@@ -368,9 +368,13 @@ cdef class _Timestamp(ABCTimestamp):
                 self = type(other)(self)
 
             # validate tz's
-            if not tz_compare(self.tzinfo, other.tzinfo):
-                raise TypeError("Timestamp subtraction must have the "
-                                "same timezones or no timezones")
+            naive_timezones = [
+                el.tzinfo is None or el.tzinfo.utcoffset(el) is None
+                for el in (self, other)
+            ]
+            if any(naive_timezones) and not all(naive_timezones):
+                raise TypeError("Cannot compute time difference between"
```

Comment:
I'd do: `Timedelta between Timestamps with different timezone object types is not supported.`

Reply:
could use assert_tzawareness_compat

Reply:
i don't think the suggested error message is precise enough, the key issue here is whether you can ascertain the timezone of both dates with confidence:

* if both are naive, we can assume they are both in local time; the time difference doesn't depend on the timezone
* if both are timezone-aware, we can compute the absolute time difference (e.g. 12pm EST - 12PM PT is 3 hrs; 12pm EST - 9AM PT is 0 hrs)
* if only one is naive, we have to fail because we can't know what the timezone offset is between the naive and non-naive times

This also makes me think that we should be raising `ValueError` here instead of `TypeError`.